python/lsst/ts/ataos/ataos_csc.py

```diff
@@ -782,78 +993,92 @@ async def correction_loop(self):
+                                self.log.debug(f"Adding {correction} correction.")
+                                corrections_to_apply.append(
+                                    self.corrections_routines[correction](azimuth,
+                                                                          elevation))
+
+                        # Check to see if any model offsets need applying due
+                        # to changes in atspectrograph filter/disperser
+                        # changes, then apply if needed
+                        self.log.debug("self.corrections['atspectrograph'] "
+                                       f"is {self.corrections['atspectrograph']}")
+                        self.log.debug("self.focus_offset_yet_to_be_applied "
+                                       f"is {self.focus_offset_yet_to_be_applied}")
+                        self.log.debug("self.pointing_offsets_yet_to_be_applied "
+                                       f"is {self.pointing_offsets_yet_to_be_applied}")
+
+                    else:
+                        # This should not raise an exception as it would
+                        # require the ATMCS is online for the ATAOS to
+                        # be online. Having the loop running (but just
+                        # sleeping) is acceptable and allows cycling of
+                        # components.
+                        self.log.debug("No information available about telescope azimuth and/or "
+                                       "elevation, m1, m2, hexapod corrections cannot occur ")
+
+                    # FIXME:
+                    # Run corrections in series because CSCs are not
+                    # supporting concurrent operations yet 2019/June/4
+
+                    for corr in corrections_to_apply:
+                        await corr
+
+                    # run corrections concurrently (and/or wait for the
+                    # heartbeat interval)
+                    # if len(corrections_to_apply) > 0:
+                    #     await asyncio.gather(*corrections_to_apply)
+
+                except asyncio.CancelledError:
+                    self.log.debug("Correction loop cancelled.")
+                    break
+                except Exception:
+                    self.log.exception("Error in correction loop. Going to FAULT state.")
+                    self.fault(code=CORRECTION_LOOP_DIED,
+                               report="Correction loop died.",
+                               traceback=traceback.format_exc())
+                    self.fault()
+                    break
+
+            # Lock releases when outside the "with" statement
+            # Post asyncio.Event (not a DDS Event) saying the loop has
+            # finished and lock is released
+            self.correction_loop_completed_evt.set()
+            # wait the loop time before next iteration
+            await asyncio.sleep(self.correction_loop_time)
```

Comment:
So if we leave the sleep in there then the lock is being held while the sleep is on-going which I don't think is what we want to do.
I do agree that having it outside the loop now makes the loop not run at 1Hz. One way we could get around this is to time how long the correction takes, then only sleep for that amount of time.
However, this could create other issues. If the correction takes longer than 1s, then you could get two subsequent corrections and I'm not sure if the queue would start filling up.
Of course, one could work around that as well, requiring a ~0.1s minimum sleep after the loop.
I suppose the moral of the story is that there is no way guarantee the loop runs at a fixed rate... but do we care? I suspect not.
For the moment I'm removing the comment on line 1009 but if we want to change how this loop timing/implemention is done I'm happy to have that conversation.